xrootd/xrdproto/stat/stat.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2018 The go-hep Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package stat contains the structures describing request and response for stat request.
+// See xrootd protocol specification (http://xrootd.org/doc/dev45/XRdv310.pdf, p. 113) for details.
+package stat // import "go-hep.org/x/hep/xrootd/xrdproto/stat"
+
+import (
+	"go-hep.org/x/hep/xrootd/internal/xrdenc"
+	"go-hep.org/x/hep/xrootd/xrdfs"
+)
+
+// RequestID is the id of the request, it is sent as part of message.
+// See xrootd protocol specification for details: http://xrootd.org/doc/dev45/XRdv310.pdf, 2.3 Client Request Format.
+const RequestID uint16 = 3017
+
+// DefaultResponse is a response for the stat request which contains stat information such as:
+// the OS-dependent identifier, the size of the data, the entry attributes and the modification time.
+type DefaultResponse struct {
+	xrdfs.EntryStat
+}
+
+// VirtualFSResponse is a response for the stat request
+// which contains virtual file system stat information such as:
+// nrw - the number of nodes that can provide read/write access,
+// frw - the size of the largest contiguous area of r/w free space,
+// urw - the percent utilization of the partition represented by frw,
+// nstg - the number of nodes that can provide staging access,
+// fstg - the size of the largest contiguous area of staging free space,
+// ustg - the percent utilization of the partition represebted by fstg,
+type VirtualFSResponse struct {
+	xrdfs.VirtualFSStat
+}
+
+// RespID implements xrdproto.Response.RespID
+func (resp *VirtualFSResponse) RespID() uint16 { return RequestID }
+
+// RespID implements xrdproto.Response.RespID
+func (resp *DefaultResponse) RespID() uint16 { return RequestID }
+
+// Options are stat processing options.
+type Options uint8
+
+const (
+	OptionsVFS Options = 1 // OptionsVFS indicates that virtual file system information is requested.
+)
+
+// Request holds open request parameters.
+type Request struct {
+	Options    Options
+	_          [11]uint8
+	FileHandle xrdfs.FileHandle
+	Path       string
+}
+
+// NewRequestForPath forms a Request according to provided path
+// and virtualFS parameter which indicates whether virtual file system information
+// should be returned instead of default stat information.
+func NewRequestForPath(path string, virtualFS bool) *Request {
```

Comment:
not sure this function (nor the one below: `NewRequestForHandle`) actually improves the usability.

consider:
```go
req := stat.NewRequestForPath("/foo/bar", false)
req := stat.Request{Path: "/foo/bar"}
```
or:
```go
req := stat.NewRequestForPath("/foo/bar", true)
req := stat.Request{Path: "/foo/bar", Options: stat.OptionsVFS}
```

IMHO, it reads better using the named-struct-fields way :)

I would have probably gone with the `NewXyz` functions for forward compatibility in the event that `stat.Request` gained new `Options` or new fields (recouped from the `[11]byte` reserve), but the specs require that everything is zeroed, so new versions of `Request` would have to differ.

what do you think?

Reply:
> but the specs require that everything is zeroed

I got it this way: in the future, these zeroed values can be used for passing some values that will have zero as their default value.

Let's assume that some of the reserved values is changed to the meaningful parameter. Probably the default value (`0`) will be suitable for the most of cases, so backward compatibility will be kept.

I think that in most of cases we will not want to change these default values (to keep backward compatibility from the our side too). So, `NewRequestForPath` will not be modified and another func will be introduced.

From that point of view it's ok to use `Request{ ... }` instead of `NewRequestForXyz`.